trieste/data.py

```diff
@@ -108,3 +109,90 @@ def astuple(self) -> tuple[TensorType, TensorType]:
         :return: A 2-tuple of the :attr:`query_points` and :attr:`observations`.
         """
         return self.query_points, self.observations
+
+
+def check_and_extract_fidelity_query_points(
+    query_points: TensorType, max_fidelity: Optional[int] = None
+) -> tuple[TensorType, TensorType]:
+    """Check whether the final column of a tensor is close enough to ints
+    to be reasonably considered to represent fidelities.
+
+    The final input column of multi-fidelity data should be a reference to
+    the fidelity of the query point. We cannot have mixed type tensors, but
+    we can check that thhe final column values are suitably close to integers.
+
+    :param query_points: Data to check final column of.
+    :raise: ValueError: If there are not enough columns to be multifidelity data
+    :raise InvalidArgumentError: If any value in the final column is far from an integer
+    :return: Query points without fidelity column
+        and the fidelities of each of the query points
+    """
+    # Check we have sufficient columns
+    if query_points.shape[1] < 2:
```

Comment:
here and below: it might be useful to assume that the querypoints can be higher order tensors? eg
`got {query_points.shape[-1]}`
`input_points = query_points[..., :-1]`
etc.